hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java

```diff
@@ -90,6 +117,79 @@ void setScheduledFuture(ScheduledFuture<?> future) {
         this.scheduledFuture = future;
     }
 
+    void setResult(Object result) {
+        this.result.set(result);
+    }
+
+    Object getResult()
+            throws ExecutionException, InterruptedException {
+
+        return unpeelFutureOrRoot().get();
+    }
+
+    boolean cancel(boolean mayInterrupt)
+            throws ExecutionException, InterruptedException {
+
+        if (!cancelled.compareAndSet(false, true) || scheduledFuture == null) {
+            return false;
+        }
+
+        // Unpeel Executor's Future (in case of Callable, see. DelegatingCallableTaskDecorator)
+        Future delegatedFuture = null;
```

Comment:
Also, this approach has the problem that for a single-run callable to get cancelled, it will first wait for it to get scheduled before cancelling it. Here is one possible approach to fix this :

```
final boolean cancelled = scheduledFuture.cancel(mayInterrupt);
        if (SINGLE_RUN.equals(definition.getType())) {
            try {
                return ((Future) scheduledFuture.get()).cancel(mayInterrupt);
            } catch (CancellationException e) {
                // SCHEDULING WAS ALREADY CANCELLED, ABORT
            }
        }
        return cancelled;
```